### PR TITLE
RBAC: Add an endpoint to list all user permissions

### PR DESCRIPTION
**Pull request by gamab**
Co-authored-by: Joey Orlando

**What is this feature?**

RBAC: Add an endpoint to list all user permissions

**Why do we need this feature?**

We are currently covering OnCall with RBAC. OnCall used to sync the organization roles from Grafana to their database. They want to reproduce the same behavior at first.
This endpoint will allow them to do so.

Note that this endpoint doesn't scale, it has very bad performances when reaching ~5k users and ~5k plugin related permissions per user.

OnCall team is going to look on a way to sync permissions per user when needed but still needs this endpoint for specific use cases like "who has the necessary permission to be notified"

This endpoint make sense for debugging and monitoring purpose.

This work is behind a feature toggle, if we don't manage to make it scale, the endpoint will be removed with the feature-toggle.

**Who is this feature for?**

OnCall

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana-enterprise/issues/4053

**Special notes for your reviewer**:

ENT counterpart: https://github.com/grafana/grafana-enterprise/pull/4054

### CODE REVIEW

**Comment by gamab** [comment]
On hold, we might be able to work with a per user endpoint in the end :tada:

File: pkg/services/accesscontrol/accesscontrol.go [hunk 1]
```diff
@@ -47,6 +49,12 @@ type Options struct {
 	ReloadCache bool
 }
 
+type SearchOptions struct {
+	ActionPrefix string
+	Action       string
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/accesscontrol.go:54]: thought: could we just have `ActionPrefix` as an option, and if the user wants to filter by a specific action then they pass the full action in there? The chances of one action being a prefix to another one are pretty slim, no?

File: pkg/services/accesscontrol/acimpl/service.go [hunk 2]
```diff
@@ -244,3 +248,102 @@ func (s *Service) DeclarePluginRoles(_ context.Context, ID, name string, regs []
...
+		return nil, err
+	}
+
+	// Get managed permissions (DB)
+	usersPermissions, err := s.store.SearchUsersPermissions(ctx, orgID, options)
+	if err != nil {
+		return nil, err
+	}
+
+	// helper to filter out permissions the signed in users cannot see
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/acimpl/service.go:283]: thought: should we request that the user is allowed to view all user's permissions to access this endpoint? It feels potentially dangerous to display partial permissions 🤔

File: pkg/services/accesscontrol/acimpl/service.go [hunk 3]
```diff
@@ -244,3 +248,102 @@ func (s *Service) DeclarePluginRoles(_ context.Context, ID, name string, regs []
...
+			if err != nil {
+				continue
+			}
+			ids[id] = true
+		}
+
+		return func(userID int64) bool { return ids[userID] }
+	}()
+
+	// Merge stored (DB) and basic role permissions (RAM)
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/acimpl/service.go:313]: thought: could this loop be merged with the next loop (iterate over `usersRoles` and check if `usersPermissions[userID]` is set).

File: pkg/services/accesscontrol/database/database.go [hunk 4]
```diff
@@ -55,6 +55,110 @@ func (s *AccessControlStore) GetUserPermissions(ctx context.Context, query acces
 	return result, err
 }
 
+// SearchUsersPermissions returns the list of user permissions indexed by UserID
```
Comment (IevaVasiljeva) [inline pkg/services/accesscontrol/database/database.go:58]: something to discuss: should the DB search be limited to permissions from managed roles for OSS?

**Review by IevaVasiljeva — COMMENTED** [review]
Some initial comments so that I don't lose them :)

File: pkg/services/accesscontrol/accesscontrol.go (hunk 1 above)
Comment (gamab) [reply pkg/services/accesscontrol/accesscontrol.go]: As we discussed, I'm thinking that we'll remove the `actionPrefix` possibility altogether once the PoC doesn't need it, this is why I'd prefer keeping both.

File: pkg/services/accesscontrol/acimpl/service.go (hunk 2 above)
Comment (gamab) [reply pkg/services/accesscontrol/acimpl/service.go:283]: I'm going to keep this to be consistent with other endpoints, like `/api/org/users/search`

File: pkg/services/accesscontrol/acimpl/service.go (hunk 3 above)
Comment (gamab) [reply pkg/services/accesscontrol/acimpl/service.go:313]: Done.

File: pkg/services/accesscontrol/database/database.go (hunk 4 above)
Comment (gamab) [reply pkg/services/accesscontrol/database/database.go:58]: It probably should. Let's open an issue on this and deal with it in another PR.

**Review by IevaVasiljeva — APPROVED** [review]
Amazing work and persistence on this Gabriel!

I've added some more comments (the main ones are about how scope is checked), but those can also be addressed in a follow-up work.

As discussed, I'm approving this PR to unblock the work for RBAC permissions for On-Call. We're all aware of the limitations of this solution (scalability issues), but as long as it's behind a feature flag I think it's a good stepping stone for further iterations.

File: pkg/services/accesscontrol/roles.go [hunk 5]
```diff
@@ -117,6 +117,10 @@ var (
 				Action: ActionUsersQuotasList,
 				Scope:  ScopeGlobalUsersAll,
 			},
+			{
+				Action: ActionUsersPermissionsRead,
+				Scope:  ScopeUsersAll,
+			},
```
[the remark this answers is not in the document]
Comment (gamab) [reply pkg/services/accesscontrol/roles.go]: Done. :+1:

File: pkg/services/accesscontrol/acimpl/service.go [hunk 6]
```diff
@@ -244,3 +248,90 @@ func (s *Service) DeclarePluginRoles(_ context.Context, ID, name string, regs []
...
 	return nil
 }
+
+// SearchUsersPermissions returns all users' permissions filtered by action prefixes
+func (s *Service) SearchUsersPermissions(ctx context.Context, user *user.SignedInUser, orgID int64,
+	options accesscontrol.SearchOptions) (map[int64][]accesscontrol.Permission, error) {
+	// Filter ram permissions
+	basicPermissions := map[string][]accesscontrol.Permission{}
+	for role, basicRole := range s.roles {
+		for i := range basicRole.Permissions {
```
[the remark this answers is not in the document]
Comment (gamab) [reply pkg/services/accesscontrol/acimpl/service.go:258]: Opened an issue referring to this comment. :+1:

File: pkg/services/accesscontrol/database/database.go [hunk 7]
```diff
@@ -55,6 +55,110 @@ func (s *AccessControlStore) GetUserPermissions(ctx context.Context, query acces
...
+			params = append(params, options.ActionPrefix+"%")
+		}
+		if options.Action != "" {
+			q += ` AND action = ?`
+			params = append(params, options.Action)
+		}
+		if options.Scope != "" {
+			q += ` AND scope = ?`
+			params = append(params, options.Scope)
+		}
```
[the remark this answers is not in the document]
Comment (gamab) [reply pkg/services/accesscontrol/database/database.go:112]: Opened an issue referring to this comment. :+1:

**Pull request merged**